core/src/token_info/hardcoded.rs

```diff
@@ -0,0 +1,115 @@
+//! This module contains fallback token data that should be used by the price
+//! estimator when prices are not available.
+
+use crate::models::TokenId;
+use anyhow::{anyhow, Context, Error, Result};
+
+use serde::Deserialize;
+use std::collections::HashMap;
+use std::iter::FromIterator;
+use std::str::FromStr;
+
+use super::{TokenBaseInfo, TokenInfoFetching};
+
+/// Token fallback data containing all fallback information for tokens that
+/// should be provided to the solver.
+#[cfg_attr(test, derive(Eq, PartialEq))]
+#[derive(Clone, Debug, Default, Deserialize)]
+#[serde(transparent)]
+pub struct TokenData(HashMap<TokenId, TokenBaseInfo>);
+
+impl TokenInfoFetching for TokenData {
+    fn get_token_info(&self, id: TokenId) -> Result<TokenBaseInfo> {
+        self.0
+            .get(&id)
+            .cloned()
+            .ok_or_else(|| anyhow!("Token {:?} not found in hardcoded data", id))
+    }
+
+    fn all_ids(&self) -> Result<Vec<TokenId>> {
+        Ok(Vec::from_iter(self.0.keys().copied()))
+    }
+}
+
+impl From<HashMap<TokenId, TokenBaseInfo>> for TokenData {
+    fn from(tokens: HashMap<TokenId, TokenBaseInfo>) -> Self {
+        TokenData(tokens)
+    }
+}
+
+impl FromStr for TokenData {
+    type Err = Error;
+
+    fn from_str(token_data: &str) -> Result<Self> {
+        Ok(serde_json::from_str(token_data)
+            .context("failed to parse token data from JSON string")?)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn token_fallback_data_from_str() {
+        let json = r#"{
+          "T0001": {
+            "alias": "WETH",
+            "decimals": 18,
+            "externalPrice": 200000000000000000000
+          },
+          "T0004": {
+            "alias": "USDC",
+            "decimals": 6,
+            "externalPrice": 1000000000000000000000000000000,
+            "shouldEstimatePrice": true
+          }
+        }"#;
+
+        assert_eq!(
+            TokenData::from_str(json).unwrap(),
+            TokenData::from(hash_map! {
+                TokenId(1) => TokenBaseInfo::new("WETH", 18, 200_000_000_000_000_000_000),
+                TokenId(4) => TokenBaseInfo::new("USDC", 6, 1_000_000_000_000_000_000_000_000_000_000),
+            })
+        );
+    }
+
+    #[test]
+    fn token_get_price() {
+        for (token, usd_price, expected) in &[
+            (
+                TokenBaseInfo::new("USDC", 6, 0),
+                0.99,
+                0.99 * 10f64.powi(30),
+            ),
+            (
+                TokenBaseInfo::new("DAI", 18, 0),
+                1.01,
+                1.01 * 10f64.powi(18),
+            ),
+            (TokenBaseInfo::new("FAKE", 32, 0), 1.0, 10f64.powi(4)),
+            (
+                TokenBaseInfo::new("SCAM", 42, 0),
+                10f64.powi(10),
+                10f64.powi(4),
```

Comment:
`36 - 42 + 10 = 4` check.


What happens when this results in a negative exponent?

Reply:
These tests were just moved from the original location. Ammending them is not really in the scope of this PR. A negative exponent would be a number <1, right? This is checking amounts, which are expressed as ints as far as I know.